generate/names.go

```diff
@@ -0,0 +1,157 @@
+//			... on U { h { h3 h4 } }
+//		}
+//	}
+// The field g must have type QFIG (not QFTG and QFHG), so that QFI's method
+// GetG() can return a consistent type.  But the fields h must have types QFTH
+// and QFUH (not QFIH), because the two are different: the former has fields h1
+// and h2, whereas the latter has fields h3 and h4.  So, in summary, since `g`
+// is selected in a context of type I, it uses that (interface) type in its
+// type-name, and `h` is selected in contexts of types T and U, they use those
+// (implementation) types in their type-names.
+//
+// We do shorten the names in one case: if the name of a field ends with
+// the name of its type, we omit the type name, avoiding types like
+// MyQueryUserUser when querying a field of type user and value user.  Note we
+// do not do this for field names, both because it's less common, and because
+// in `query Q { user { user { id } } }` we do need both QUser and QUserUser --
+// they have different fields.
+//
+// Note that there are (at least) two potential collisions from this algorithm:
+// - When generating Go types for GraphQL interface types, we generate both
+//   ...MyFieldMyInterfaceType and ...MyFieldMyImplType.  If an interface's
+//   name is a suffix of its implementation's name, and both are suffixes of a
+//   field of that type, we'll shorten both, resulting in a collision.
+// - Given a query like
+//		query Q {
+//			ab { ... } # type: C
+//			a { ... }  # type: BC
+//		}
+//	 we generate QABC to represent both fields.
+// Both cases seem fairly rare in practice; eventually we'll likely allow users
+// the ability to specify their own names, which they could use to avoid this
+// (see https://github.com/Khan/genqlient/issues/12).
+// TODO(benkraft): We should probably at least try to detect it and bail.
+//
+// To implement all of the above, as we traverse the operation (and schema) in
+// convert.go, we keep track of a list of parts to prefix to our type-names.
+// The list always ends with a field, not a type; and we extend it when
+// traversing fields, to allow for correct handling of the interface case
+// discussed above.  This file implements the actual maintenance of that
+// prefix, and the code to compute the actual type-name from it.
+//
+// Note that input objects and enums are handled separately (inline in
+// convertDefinition) since the same considerations don't apply and their names
+// are thus quite simple.  We also specially-handle the type of the toplevel
+// response object (inline in convertOperation).
+
+import (
+	"strings"
+
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+// Yes, a linked list!  We could use a stack -- it would probably be marginally
+// more efficient -- but then the caller would have to know more about how to
+// manage it safely.  Using a list, and treating it as immutable, makes it
+// easy.
+type prefixList struct {
+	last string // the list goes back-to-front, so this is the *last* prefix
+	rest *prefixList
```

Comment:
nit: perhaps `previous` instead of `rest`? Or maybe name `last` either `prefix` or `name` or `element` and say this is a forward linked list but with the elements in the opposite order?

Reply:
Ah, good ideas.